### DISCUSSION TITLE
Question about FHIR search parameters again

### DISCUSSION THREAD

**Discussion by m-onyshchuk** [LinuxForHealth/connect#354] [Q&A] [answered]
I have saved Patient resource with favorite-color in extension:
```
{
  "resourceType":"Patient",
  "id":"183800c30d7-12001484-9317-416f-9194-2edad26dbd38",
  ...
  "extension":[
    {
      "url":"http://example.com/fhir/extension/Patient/favorite-color",
      "valueIdentifier":{
        "system":"http://example.com/fhir/extension/Patient/favorite-color",
        "value":"black"
      }
    }
  ...  
  ]
}
```
My `/opt/ol/wlp/usr/servers/defaultServer/config/default/extension-search-parameters.json` file is:

```
{
  "resourceType": "Bundle",
  "type": "collection",
  "entry": [
    {
      "fullUrl": "http://example.com/fhir/SearchParameter/Patient-favorite-color",
      "resource": {
        "resourceType": "SearchParameter",
        "id": "Patient-favorite-color",
        "url": "http://example.com/fhir/SearchParameter/Patient-favorite-color",
        "version": "4.0.0",
        "name": "favorite-color",
        "status": "draft",
        "experimental": false,
        "date": "2018-12-27T22:37:54+11:00",
        "publisher": "LinuxForHealth FHIR Server Test",
        "contact": [{
          "telecom": [{
            "system": "url",
            "value": "http://linuxforhealth.org"
          }]
        },
          {
            "telecom": [{
              "system": "url",
              "value": "http://linuxforhealth.org"
            }]
          }],
        "description": "the patient's favorite color",
        "code": "favorite-color",
        "base": ["Patient"],
        "type": "string",
        "xpathUsage": "normal",
        "xpath": "f:Patient/f:extension[@url='http://example.com/fhir/extension/Patient/favorite-color']/f:valueString",
        "expression": "Patient.extension.where(url='http://example.com/fhir/extension/Patient/favorite-color').value",
        "multipleOr": true,
        "multipleAnd": true,
        "modifier": []
      }
    }
  ]
}

```
Next I have tried to find my Patient resource by http request:
```
GET /fhir-server/api/v4/Patient?favorite-color=black
```
and got the aswer:
```
{
  "resourceType": "Bundle",
  "id": "1c626a4c-4996-433d-9017-61c0f7d2ec89",
  "type": "searchset",
  "total": 0,
  "link": [
    {
      "relation": "self",
      "url": "https://localhost:9443/fhir-server/api/v4/Patient?_count=10&favorite-color=black&_page=1"
    }
  ]
}
```
What should I change to find my Patient by favorite-color? 
P.S. favorite-color is custom search parameter from example: https://linuxforhealth.github.io/FHIR/guides/FHIRSearchConfiguration#11-tenant-specific-parameters

**Comment by lmsurpre** [comment, marked as answer]
check out the FHIR search summary tables at https://www.hl7.org/fhir/search.html#table specifically the Identifier row.

<img width="690" alt="image" src="https://user-images.githubusercontent.com/5478798/192623490-68f0587c-3ba7-49df-9c56-d8e624d10990.png">

note that this data type is only indexed by search parameters of type "token" (whereas yours is a string)

**Comment by m-onyshchuk** [comment]
I'm sorry, I did not fully understand your advice on how to do a search by extensions in accordance with the example.
What do I have to do to get search by favorite-color extension to work:
1) change in Patient extension `valueIdentifier` to `valueString` ? 
2) change something in `search-bundle.json` file ?
3) something else ?

**Reply by lmsurpre** [reply]
either 1 or 2.

if you want to use the existing search parameter, change the data to use a valueString (because that type is indexable via a search parameter of type "string").

if you want the data to be an identifier, then you need to change `"type": "string"` to `"type": "token"` on the SearchParameter (tokens are usually search like `[base]/[resourceType]?param=system|value`)

**Reply by m-onyshchuk** [reply]
I chosed second version and now I can search by my extension in Patient resource, but similar search parameter in other resource (CarePlan) doesn`t work.
Command ```sudo docker logs fhir-server | grep search``` does not show any errors in search param setup.  How can I get more dignostics about serch params or how can I get all existing search parameters in IBM FHIR Server?

**Reply by lmsurpre** [reply]
> How can I get more diagnostics about search params

You can selectively enable tracing for specific packages or even specific classes by setting TRACE_SPEC via either env var or by configDropin.  https://linuxforhealth.github.io/FHIR/guides/FHIRServerUsersGuide#312-logging-and-trace

>  how can I get all existing search parameters in IBM FHIR Server?

All configured search parameters are presented in the server's CapabilityStatement when you `GET [base]/fhir-server/api/v4/metadata`
